packages/fxa-auth-server/lib/routes/subscriptions.js

```diff
@@ -10,6 +10,597 @@ const ScopeSet = require('../../../fxa-shared').oauth.scopes;
+    //   2) Paid subscription, stop and return as they already have the sub
+    //   3) Old subscription will have no open invoices, ignore it
+    if (subscription && subscription.latest_invoice) {
+      let invoice = /** @type {Invoice} */ (subscription.latest_invoice);
+      if (invoice.status === 'open') {
+        const payment_intent =
+          /** @type {PaymentIntent} */ (invoice.payment_intent);
+        if (payment_intent.status === 'requires_payment_method') {
+          // Re-run the payment
+          invoice = await this.payments.stripe.invoices.pay(invoice.id, {
+            expand: ['payment_intent'],
+          });
+          if (!this.paidInvoice(invoice)) {
+            throw error.paymentFailed();
+          }
+        } else {
+          throw error.backendServiceFailure('stripe', 'invoice status', {
+            invoiceId: invoice.id,
+            invoiceStatus: invoice.status,
+            paymentStatus: payment_intent.status,
+          });
+        }
+      } else if (invoice.status === 'paid') {
+        throw error.subscriptionAlreadyExists();
+      } else {
+        subscription = undefined;
+      }
+    }
+
+    if (!subscription) {
+      // Create the subscription
+      subscription = await this.payments.stripe.subscriptions.create({
+        customer: customer.id,
+        items: [{ plan: selectedPlan.plan_id }],
+        expand: ['latest_invoice.payment_intent'],
+      });
+
+      if (
+        !this.paidInvoice(
+          /** @type {import('stripe').Stripe.Invoice} */ (subscription.latest_invoice)
+        )
+      ) {
+        throw error.paymentFailed();
+      }
+    }
+
+    // Store the record in our local database
+    await this.db.createAccountSubscription({
+      uid,
+      subscriptionId: subscription.id,
+      productId,
+      // Stripe create is in seconds, we use milliseconds
+      createdAt: subscription.created * 1000,
+    });
+    const devices = await request.app.devices;
+    await this.push.notifyProfileUpdated(uid, devices);
+    this.log.notifyAttachedServices('profileDataChanged', request, {
+      uid,
+      email,
+    });
+    await this.profile.deleteCache(uid);
+    const account = await this.db.account(uid);
+    await this.mailer.sendDownloadSubscriptionEmail(account.emails, account, {
+      acceptLanguage: account.locale,
+      productId,
+    });
+    this.log.info('subscriptions.createSubscription.success', {
+      uid,
+      subscriptionId: subscription.id,
+    });
+    return {
+      subscriptionId: subscription.id,
+    };
+  }
+
+  /**
+   * Verify that the invoice was paid successfully.
+   *
+   * Note that the invoice *must have the `payment_intent` expanded*
+   * or this function will fail.
+   *
+   * @param {Invoice} invoice
+   * @returns {boolean}
+   */
+  paidInvoice(invoice) {
+    if (typeof invoice.payment_intent === 'string') {
+      throw error.backendServiceFailure('stripe', 'PI fetch');
```

Comment:
Maybe expand `PI` in the message.

Reply:
Ah, this is a missed type-cast, I'll fix that.